plugins/flashrom-i2c/fu-i2c-device.c

```diff
@@ -0,0 +1,288 @@
+/* Copyright 2020 The Chromium OS Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+
+#include "config.h"
+#include "fu-i2c-device.h"
+
+#include <glib.h>
+#include <glib/gstdio.h>
+
+#define LAYOUT_FLAG_NAME "FLAG"
+#define LAYOUT_PARTITION_NAME "BLK"
+#define IMG_LAYOUT_NAME "layout"
+#define IMG_FLAG1_NAME "flag1.bin"
+#define IMG_FLAG2_NAME "flag2.bin"
+#define IMG_FIRMWARE_NAME "fw.bin"
+
+struct FlashromArgs
+{
+	const gchar *spi_master;
+	const gchar *layout;
+	const gchar *image;
+	const gchar *operation;
+};
+
+struct _FuI2cDevice
+{
+	FuDevice parent_instance;
+};
+
+G_DEFINE_TYPE (FuI2cDevice, fu_i2c_device, FU_TYPE_DEVICE)
+
+/* This calls fu_common_get_contents_fd which will close the fd when done. */
+static gint fu_i2c_device_get_boot_block_from_fd (int fd, GError **error)
+{
+	g_autoptr (GBytes) result = NULL;
+	gsize result_len = 0;
+	result = fu_common_get_contents_fd (fd, 1, error);
+	if (result == NULL)
+		return -1;
+
+	const guint8 *result_bytes = g_bytes_get_data (result, &result_len);
+	if (result_len != 1) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s: bootblock info size is expected to be 1 byte, \
+			     got %d",
+			     __func__,
+			     result_len);
+		return -1;
+	}
+
+	return result_bytes[0];
+}
+
+static gboolean fu_i2c_device_validate_flashrom_args (
+	const struct FlashromArgs *args,
+	GError **error)
+{
+	if (args->spi_master == NULL ||
+	    args->layout == NULL ||
+	    args->image == NULL ||
+	    args->operation == NULL) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "%s: all arguments under FlashromArgs has \
+			     to be set",
+			     __func__);
+		return FALSE;
+	}
+
+	return TRUE;
+
+}
+
+static gboolean fu_i2c_device_run_command (const struct FlashromArgs *args,
+					   GError **error)
+{
+	if (fu_i2c_device_validate_flashrom_args (args, error) == FALSE)
+		return FALSE;
+
+	const gchar *argv[9] = { "flashrom",
+				 "-p", args->spi_master,
+				 "--layout", args->layout,
+				 "--image", args->image,
+				 args->operation,
+				 NULL };
+
+	return fu_common_spawn_sync (argv, NULL, NULL, 0, NULL, error);
+}
+
+/* Update should happen on the non-activated block. If neither block is
+ * activated, force update to block 1.
+ */
+static gint fu_i2c_device_get_target_block_no (
+	const gchar *dir_name,
+	const gchar *spi_master,
+	const gchar *layout,
+	const gchar *flag_name,
+	GError **error)
+{
+	struct FlashromArgs flash_args;
+	g_autofree gchar *img_arg = NULL;
+	g_autofree gchar *tmp_file_name = NULL;
+	tmp_file_name = g_build_filename (dir_name, "XXXXXX", NULL);
+	gint fd = g_mkstemp (tmp_file_name);
```

Comment:
I don't think we want to be creating temp files.

Reply:
Is there any particular reason for that? If it's just no temp file, we can just create a normal file inside the temp dir.

Reply:
So specifically I think the temp file is an artifact of interacting with another executable binary, which as mentioned in other comments is not directionally where we want to be going with fwupd.  Where possible we work directly with sysfs interfaces or via ioctl on devfs interfaces.  If there is an abstraction library, then we use that.  So `libflashrom` is the intended way to work with devices that otherwise use flashrom for flashing.